src/client/activation/migrateDataScienceSettingsService.ts

```diff
@@ -48,37 +74,88 @@ export class MigrateDataScienceSettingsService implements IExtensionActivationSe
+                // Special case. URI is no longer supported. Move it to storage
                 this.serverUriStorage.setUri(val).ignoreErrors();
+                newKey = 'jupyter.jupyterServerType';
 
                 // Set the setting to local or remote based on if URI is 'local'
                 val = val === Settings.JupyterServerLocalLaunch ? val : Settings.JupyterServerRemoteLaunch;
-
-                // Change the key to the jupyter version (still needs the 19 chars in front so it substr correctly)
-                k = 'xxxxxx.dataScience.jupyterServerType';
+            } else if (!content.hasOwnProperty(newKey)) {
+                if (typeof val === 'string') {
+                    // If the value contains references to python.dataScience.* commands, migrate those too
+                    // There may be multiple occurrences of commands in the object value
+                    val = val.replace(/python\.dataScience\./gi, 'jupyter.');
+                }
             }
 
             // Update the new value
-            fileContents = applyEdits(
-                fileContents,
-                modify(fileContents, [`jupyter.${k.substr(19)}`], val, modificationOptions)
-            );
+            fileContents = applyEdits(fileContents, modify(fileContents, [newKey], val, modificationOptions));
         });
 
         await this.fs.writeLocalFile(filePath, fileContents);
-        return fileContents;
+    }
+
+    // Users may have mapped old python.dataScience.* commands to custom keybindings
+    // in their user keybindings.json. Ensure we migrate these too.
+    private async fixKeybindingsFile(filePath: string) {
+        const fileContents = await this.fs.readLocalFile(filePath);
+        if (!fileContents.includes('python.dataScience.')) {
+            return;
+        }
+        const errors: ParseError[] = [];
+        const keybindings = parse(fileContents, errors, { allowTrailingComma: true, disallowComments: false });
+        if (errors.length > 0) {
+            traceError('JSONC parser returned ParseError codes', errors);
+            return;
+        }
+        if (!Array.isArray(keybindings)) {
+            return;
+        }
+        // Possible the user already migrated the setting themselves
+        // Ensure we don't migrate a setting if the replacement command already exists
+        const jupyterCommands = keybindings.map((keybinding) => {
+            const command = keybinding.command;
+            if (typeof command === 'string' && command.startsWith('jupyter.')) {
+                return command;
+            }
+        });
+        const migratedKeybindings: IKeyBinding[] = [];
+        keybindings.forEach((keybinding) => {
+            const command = keybinding.command;
+            if (typeof command === 'string' && command.startsWith('python.dataScience.')) {
+                const targetCommand = `jupyter.${command.substr(19)}`;
+                if (!jupyterCommands.includes(targetCommand)) {
```

Comment:
@DonJayamanne here's the code for checking for dupes. First collect all the jupyter commands, then when migrating the python.dataScience commands, only include them if the equivalent jupyter command isn't mapped to something in the keybindings.json